python_transport/wirepas_gateway/transport_service.py

```diff
@@ -27,6 +27,127 @@
+
+    It could also be implemented without a periodic check and tested
+    each time a packet is published but on performance point of view it would
+    be the same
+    """
+    def __init__(self, logger, period, mqtt_wrapper, sink_manager,
+                 minimum_sink_cost, max_buffered_packets, max_delay_without_publish):
+        Thread.__init__(self)
+
+        self.logger = logger
+        # Daemonize thread to exit with full process
+        self.daemon = True
+
+        # How often to check the queue
+        self.period = period
+        self.mqtt_wrapper = mqtt_wrapper
+        self.sink_manager = sink_manager
+
+        self.running = False
+        self.disconnected = False
+
+        # Get parameters for black hole algorithm detection
+        self.minimum_sink_cost = minimum_sink_cost
+        self.max_buffered_packets = max_buffered_packets
+        self.max_delay_without_publish = max_delay_without_publish
+
+    def _set_sinks_cost(self, cost):
+        for sink in self.sink_manager.get_sinks():
+            sink.cost = cost
+
+    def _set_sinks_cost_high(self):
+        self._set_sinks_cost(self.SINK_COST_HIGH)
+
+    def _set_sinks_cost_low(self):
+        self._set_sinks_cost(self.minimum_sink_cost)
+
+    def _is_publish_delay_over(self):
+        if self.max_delay_without_publish <= 0:
+            # No max delay set, not enabled
+            return False
+
+        if self.mqtt_wrapper.publish_gueue_size <= 0:
+            # No packet queued, so nothing to check
+            return False
+
+        return self.mqtt_wrapper.last_published_packet_s > self.max_delay_without_publish
+
+    def _is_buffer_threshold_reached(self):
+        if self.max_buffered_packets <= 0:
+            # Useless check as mechanism is disabled if max is 0
+            return False
+
+        return self.mqtt_wrapper.publish_gueue_size > self.max_buffered_packets
+
+    def run(self):
+        """
+        Main loop that check periodically the status of the published queue and compare
+        it to threshold set when starting Transport
+        """
+
+        # Initialize already detected sinks
+        self._set_sinks_cost_low()
+
+        self.running = True
+
+        while self.running:
+            if not self.disconnected:
+                # Check if a condition to declare "back hole" is met
+                if self._is_publish_delay_over() or self._is_buffer_threshold_reached():
+                    self.logger.info("Increasing sink cost of all sinks")
+                    self.logger.debug("Last publish: %s Queue Size %s",
+                                      self.mqtt_wrapper.last_published_packet_s,
+                                      self.mqtt_wrapper.publish_gueue_size)
+
+                    self._set_sinks_cost_high()
+                    self.disconnected = True
+            else:
+                if self.mqtt_wrapper.publish_gueue_size == 0:
+                    # Network is back, put the connection back
```

Comment:
I think it could take a while for the buffer queue to reach zero, especially if it has been queuing up packets for a while. Aslo, could we end up in a race condition where the queue exhaust is not sufficient to compensate the consumption of network packets?

Reply:
I'm not sure to get your point.
Yes it may take a while (depending on the maximum number of packets set), and that's why I prefer to flush everything before reducing the sink cost again.
I don't really understand the race you mentioned. The thread check periodically the status of queue so at some point the queue will be empty.

Reply:
I was aiming at a situation where we try to fill and exhaust the queue at the same time, thus we could end up in a situation where we could not reach zero. But if you guarantee that the queue is exhausted first, then it is okay :+1: